File: services/app/internal/handlers/category/save_category.go
package category

import (
	"encoding/json"
	"net/http"

	"github.com/TimoSto/ThesorTeX/backend/pkg/log"
	bib_categories2 "github.com/TimoSto/ThesorTeX/services/app/internal/bib_categories"
	"github.com/TimoSto/ThesorTeX/services/app/internal/database"
)

type SaveData struct {
	Project        string
	Name           string
	InitialName    string
	CitaviCategory string
	CitaviFilter   []string
	BibFields      []bib_categories2.Field
	CiteFields     []bib_categories2.Field
}

func HandleSaveCategory(store database.ThesorTeXStore) http.Handler {
	fn := func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPut {
			w.WriteHeader(http.StatusBadRequest)
			return
		}

		var data SaveData
		decoder := json.NewDecoder(r.Body)
		err := decoder.Decode(&data)
		if err != nil {
			log.Error("got error reading data of category save: %v", err)
			w.WriteHeader(http.StatusBadRequest)
		}

		metaData, err := bib_categories2.SaveCategory(store, data.Project, data.Name, data.InitialName, data.CitaviCategory, data.CitaviFilter, data.BibFields, data.CiteFields)
		if err != nil {
			log.Error("got error saving the category: %v", err)
			w.WriteHeader(http.StatusBadRequest)
		}

		respData, err := json.Marshal(metaData)
		if err != nil {
			log.Error("Error marshaling project meta data: %v", err)
			w.WriteHeader(http.StatusInternalServerError)
		}

		w.Write(respData)
	}

	return http.HandlerFunc(fn)
}

File: services/app/frontend/src/api/projectData/categories/SaveCategory.ts
import {Field} from "./BibCategory";
import SaveResponse from "../SaveResponse";

interface CategorySaveData {
  Project: string
  Name: string
  InitialName: string,
  CitaviCategory: string,
  CitaviFilter: string[],
  BibFields: Field[],
  CiteFields: Field[]

}
export default async function SaveCategory(project: string, name: string, initialName: string, citaviCategory: string, citaviFilter: string[], bibFields: Field[], citeFields: Field[]): Promise<SaveResponse> {
  const obj: CategorySaveData = {
    Project: project,
    InitialName: initialName,
    Name: name,
    CitaviCategory: citaviCategory,
    CitaviFilter: citaviFilter,
    BibFields: bibFields,
    CiteFields: citeFields
  }

  const data = JSON.stringify(obj)

  const resp = await fetch('/app/saveCategory', {
    method: 'PUT',
    body: data
  });


  const respData = await resp.json();

  return {
    Ok: resp.ok,
    Data: respData
  };
}
